src/tree.ts

```diff
@@ -57,52 +57,201 @@ export class Tree {
     return this.rootNode.root;
   }
 
-  getNode(index: Gindex): Node {
+  getNode(gindex: Gindex): Node {
     let node = this.rootNode;
-    for (const i of gindexIterator(index)) {
-      if (i) {
-        if (node.isLeaf()) throw new Error(ERR_INVALID_TREE);
+    if (gindex < 1) {
+      throw new Error("Invalid gindex < 1");
+    }
+    const bitstring = gindex.toString(2);
+    for (let i = 1; i < bitstring.length - 1; i++) {
+      node = bitstring[i] === "0" ? node.left : node.right;
+    }
+    return node;
+  }
+
+  setNode(gindex: Gindex, n: Node, expand = false): void {
+    let node = this.rootNode;
+
+    // Pre-compute entire bitstring instead of using an iterator (25% faster)
+    if (gindex < 1) {
+      throw new Error("Invalid gindex < 1");
+    }
+    const bitstring = gindex.toString(2);
+
+    // Keep a list of all parent nodes of node at gindex `index`. Then walk the list
+    // backwards to rebind them "recursively" with the new nodes without using functions
+    const parentNodes: Node[] = [this.rootNode];
+
+    // Ignore the first bit, left right directions are at bits [1,..]
+    // Ignore the last bit, no need to push the target node to the parentNodes array
+    for (let i = 1; i < bitstring.length - 1; i++) {
+      if (node.isLeaf()) {
+        if (!expand) {
+          throw new Error(ERR_INVALID_TREE);
+        } else {
+          node = zeroNode(bitstring.length - i);
+        }
+      }
+
+      // Compare to string directly to prevent unnecessary type conversions
+      if (bitstring[i] === "1") {
         node = node.right;
       } else {
-        if (node.isLeaf()) throw new Error(ERR_INVALID_TREE);
         node = node.left;
       }
+
+      parentNodes.push(node);
     }
-    return node;
+
+    node = n;
+
+    // Ignore the first bit, left right directions are at bits [1,..]
+    // Iterate the list backwards including the last bit, but offset the parentNodes array
+    // by one since the first bit in bitstring was ignored in the previous loop
+    for (let i = bitstring.length - 1; i >= 1; i--) {
+      if (bitstring[i] === "1") {
+        node = parentNodes[i - 1].rebindRight(node);
+      } else {
+        node = parentNodes[i - 1].rebindLeft(node);
+      }
+    }
+
+    this.rootNode = node;
   }
 
-  setter(index: Gindex, expand = false): Link {
-    let link = identity;
+  /**
+   * Set multiple nodes in batch, editing and traversing nodes strictly once.
+   * gindexes MUST be sorted in ascending order beforehand. All gindexes must be
+   * at the exact same depth.
+   *
+   * Strategy: for each gindex in `gindexes` navigate to the depth of its parent,
+   * and create a new parent. Then calculate the closest common depth with the next
+   * gindex and navigate upwards creating or caching nodes as necessary. Loop and repeat.
+   */
+  setNodes(gindexes: Gindex[], nodes: Node[]): void {
```

Comment:
should this be a `Map<Gindex, Node>`?

Reply:
Maps take more memory than regular arrays. Also, it convenient to iterate both, no problem. Once consuming this we can see if it's inconvenient.

Reply:
Map may not be a bad idea. I foresee that it may be useful for building up an update when we need to update a single gindex multiple times, like when ssz vectors are tightly packed. (Increased memory shouldn't be a problem because its temporary data?)
We'll know for sure when we try to integrate this into ssz.